ale_linters/haskell/hie.vim

```diff
@@ -0,0 +1,29 @@
+" Author: Luxed <devildead13@gmail.com>
+" Description: A language server for Haskell
+
+call ale#Set('haskell_hie_executable', 'hie')
+
+function! ale_linters#haskell#hie#GetExecutable(buffer) abort
+    return ale#Var(a:buffer, 'haskell_hie_executable')
+endfunction
+
+function! ale_linters#haskell#hie#GetCommand(buffer) abort
+    let l:executable = ale_linters#haskell#hie#GetExecutable(a:buffer)
+    let l:arguments = '--lsp'
+
+    return ale#Escape(l:executable) . ' ' . l:arguments
+endfunction
+
+function! ale_linters#haskell#hie#GetProjectRoot(buffer) abort
+    let l:stack_file = ale#path#FindNearestFile(a:buffer, 'stack.yaml')
+
+    return !empty(l:stack_file) ? fnamemodify(l:stack_file, ':h') : getcwd()
```

Comment:
If you can't find a project file, return `''` and the language server won't be started. You can only use a language server connection for projects. As you say, it's worth looking for Cabal files too.

Reply:
Make sure not to use `getcwd()`. See `:help ale-development` for lots of information, including why `getcwd()` is banned.